File: PythonExample/NumPy/ex1/featureNormalize.py
import numpy as np
 
'''
function [X_norm, mu, sigma] = featureNormalize(X)
%FEATURENORMALIZE Normalizes the features in X 
%   FEATURENORMALIZE(X) returns a normalized version of X where
%   the mean value of each feature is 0 and the standard deviation
%   is 1. This is often a good preprocessing step to do when
%   working with learning algorithms.

% You need to set these values correctly
X_norm = X;
mu = zeros(1, size(X, 2));
sigma = zeros(1, size(X, 2));

% ====================== YOUR CODE HERE ======================
% Instructions: First, for each feature dimension, compute the mean
%               of the feature and subtract it from the dataset,
%               storing the mean value in mu. Next, compute the 
%               standard deviation of each feature and divide
%               each feature by it's standard deviation, storing
%               the standard deviation in sigma. 
%
%               Note that X is a matrix where each column is a 
%               feature and each row is an example. You need 
%               to perform the normalization separately for 
%               each feature. 
%
% Hint: You might find the 'mean' and 'std' functions useful.
%       

mu = mean(X);
sigma = std(X); 

X_norm = (X_norm - mu) / sigma;

% ============================================================

end
'''

 
# noinspection PyPep8Naming,PyPep8Naming
def feature_normalize(X):
    mu = np.mean(X, axis=0)
    sigma = np.std(X, axis=0)
    X_norm = (X - mu) / sigma
    return X_norm, mu, sigma


'''
def feature_normalize(X):
    X_norm = X
    mu = np.zeros((1, X.shape[1]))
    sigma = np.zeros((1, X.shape[1]))
    mu = np.mean(X, axis=0)
    sigma = np.std(X, axis=0)
    Mu_mat = np.tile(mu, (X.shape[0], 1)) # Extend mu to matrix of mx2
    Sigma_mat = np.tile(sigma, (X.shape[0], 1)) # Extend sigma to matrix of mx2
    X_norm = (X - Mu_mat) / Sigma_mat
    return X_norm, mu, sigma
'''

'''
def feature_normalize(X):

    #print(np.std( np.array([[1, 2], [3, 4]]) , axis=0))
    #print(np.mean(X, axis=0))
    n = X.shape[1]
 
    X_norm = X.astype('float32')
    mu = np.zeros(n)
    sigma = np.zeros(n)
    
    for i in range(n):
        feature = X_norm[:,i]
        mu[i] = np.mean(feature)
        sigma[i] = np.std(feature)
        X_norm[:,i] -= mu[i]
        X_norm[:,i] /= sigma[i]
    
    return X_norm, mu, sigma
'''

File: ex1_linear_regression/featureNormalize.m
function [X_norm, mu, sigma] = featureNormalize(X)
%FEATURENORMALIZE Нормализует функции в X
% FEATURENORMALIZE (X) возвращает нормализованную версию X, где
% среднее значение каждого признака равно 0 и стандартное отклонение
% равно 1. Часто это хороший шаг предварительной обработки, когда
% работает с алгоритмами обучения.

% You need to set these values correctly
X_norm = X;
mu = zeros(1, size(X, 2));
sigma = zeros(1, size(X, 2));

% ====================== YOUR CODE HERE ======================
% Инструкции: сначала для каждого измерения объекта вычислите среднее
% от функции и вычесть ее из набора данных,
% хранения среднего значения в мю. Далее вычисляем
% стандартного отклонения каждого признака и деления
% каждой функции по стандартному отклонению, хранение
% стандартное отклонение в сигме.
%
% Обратите внимание, что X является матрицей, где каждый столбец является
% особенность, и каждая строка является примером. Тебе нужно
% выполнить нормализацию отдельно для
% каждой функции.
%
% Подсказка: вам могут пригодиться функции mean и std.
%       
 
  % Нормализация на среднее
     %#{
      mu = mean(X);% mean - среднее арифметическое. [дефолт]  `mean([10 20 30 40 50]) == 30`
      sigma = std(X); %std - Стандартное отклонение `std([10 20 30 40 50]) ==  15.811`
      %X_norm = (X .- mu) ./ sigma; % Возможен NaN если нет малейшего отклонения то 0/0=NaN
      
      X_norm = (X .- mu) ./ sigma;
     %#}  
     
     
% ============================================================
    % Тоже рабочие варианты
    #{
      mu = mean(X);
      X_norm = bsxfun(@minus, X, mu);
      sigma = std(X_norm);
      X_norm = bsxfun(@rdivide, X_norm, sigma);
    #}

    % Тоже рабочие варианты
    #{
     for i = 1: size(X,2)
        X_norm( :, i) = (X ( :, i) - mu (i)) / sigma(i) ;
     end
    #}

    % Тоже рабочие варианты
    #{
      for i = 1:size(X,2)
        mu(i) = mean(X(:,i));
        X_norm(:,i) = X_norm(:,i) - mu(i);
        sigma(i) = std(X_norm(:,i));
        X_norm(:,i) = X_norm(:,i) / sigma(i);
       %X_norm(:,i) = ( X(:,i) - mu(i) ) / sigma(i);
     endfor
    #}

     % Тоже рабочие варианты
    #{
      mu = mean(X);
      X_norm = bsxfun(@minus, X, mu);
      sigma = std(X_norm);
      X_norm = bsxfun(@rdivide, X_norm, sigma);
    #}

% ============================================================

%!test
%! [Xn mu sigma] = featureNormalize([1 ; 2 ; 3]);
%! Xn_expected = [ -1 0 1 ]';
%! mu_expected = 2;
%! sigma_expected = 1;
%! assert(Xn, Xn_expected, 1);
%! assert(mu, mu_expected, 1);
%! assert(sigma, sigma_expected, 1);

end
